File: app/src/main/java/com/main/climbingdiary/database/TaskRepository.java
package com.main.climbingdiary.database;

import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.util.Log;

import com.main.climbingdiary.activities.MainActivity;
import com.main.climbingdiary.database.entities.Projekt;
import com.main.climbingdiary.database.entities.Route;
import com.main.climbingdiary.database.sql.SqlAreaSektoren;
import com.main.climbingdiary.database.sql.SqlRouten;
import com.main.climbingdiary.database.sql.SqlStatistic;
import com.main.climbingdiary.models.RouteType;

public class TaskRepository {

    private static final String TAG = "DataAdapter";

    private final SQLiteDatabase mDb;
    private static TaskRepository INSTANCE = null;

    private TaskRepository() {
        DatabaseHelper mDbHelper = DatabaseHelper.getInstance();
        try {
            mDbHelper.openDataBase();
            mDb = mDbHelper.getWritableDatabase();
        } catch (SQLException mSQLException) {
            Log.e(TAG, "open >>" + mSQLException.toString());
            throw mSQLException;
        }
    }

    public static synchronized TaskRepository getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new TaskRepository();
        }
        return INSTANCE;
    }

    public Cursor getAllRoutes() {
        return getCursor(SqlRouten.getRouteList(RouteType.ROUTE));
    }

    public Cursor getAllProjekts() {
        return getCursor(SqlRouten.getRouteList(RouteType.PROJEKT));
    }

    public Cursor getRoute(int id) {
        return getCursor(SqlRouten.getRoute(id));
    }

    public Cursor getProjekt(int id) {
        return getCursor(SqlRouten.getProjekt(id));
    }

    public Cursor getAllAreas() {
        return getCursor(SqlAreaSektoren.getAllAreas());
    }

    public Cursor getSectorListByAreaName(String name) {
        return getCursor(SqlAreaSektoren.getSectorByAreaName(name));
    }

    public Cursor getSectorIdByAreaNameAndSectorName(String sectorName, String areaName){
        return getCursor(SqlAreaSektoren.getSectorIdBySectorNameAndAreaName(sectorName,areaName));
    }

    public Cursor getAreaIdByAreaNameAndSectorName(String sectorName, String areaName){
        return getCursor(SqlAreaSektoren.getAreaIdBySectorNameAndAreaName(sectorName,areaName));
    }

    public Cursor getTopTenRoutes(int year) {
        return getCursor(SqlRouten.getTopTenRoutes(year));
    }

    public Cursor getYears(boolean filterSet) {
        return getCursor(SqlRouten.getYears(filterSet));
    }

    public Cursor getTableValues() {
        return getCursor(SqlStatistic.getTableValues());
    }

    public Cursor getBarChartValues() {
        return getCursor(SqlStatistic.getBarChartValues());
    }

    public boolean insertRoute(Route route) {
        //create the transaction
        return executeSqlTasks(SqlRouten.getInsertRouteTasks(route));
    }

    public boolean insertProjekt(Projekt projekt) {
        return executeSqlTasks(SqlRouten.getInsertProjektTasks(projekt));
    }

    public boolean updateRoute(Route route){
        return executeSqlTasks(new String[]{SqlRouten.updateRoute(route)});
    }

    public boolean deleteRoute(int id) {
        return executeSqlTasks(new String[]{SqlRouten.deleteRoute(id)});
    }

    public boolean deleteProjekt(int id) {
        return executeSqlTasks(new String[]{SqlRouten.deleteProjekt(id)});
    }

    public Cursor getCursor(String sql) {
        try {
            Cursor mCur = mDb.rawQuery(sql, null);
            if (mCur != null) {
                mCur.moveToNext();
            }
            return mCur;
        } catch (SQLException mSQLExeption) {
            Log.e(TAG, "Error >>" + mSQLExeption.toString());
            throw mSQLExeption;
        }
    }

    public boolean executeSqlTasks(String[] tasks) {
        mDb.beginTransaction();
        try {
            for (String x : tasks) {
                Log.d("Execute", x);
                mDb.execSQL(x);
            }
            mDb.setTransactionSuccessful();
            return true;
        } catch (SQLiteException exception) {
            Log.e(TAG, "Error >>" + exception.toString());
            return false;
        } finally {
            mDb.endTransaction();
        }
    }
}

File: app/src/main/java/com/main/climbingdiary/database/TaskRepository.kt
package com.main.climbingdiary.database

import android.annotation.SuppressLint
import android.content.Context
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.util.Log
import com.main.climbingdiary.activities.MainActivity
import com.main.climbingdiary.database.entities.Area
import com.main.climbingdiary.database.entities.Projekt
import com.main.climbingdiary.database.entities.Route
import com.main.climbingdiary.database.entities.Sector
import com.main.climbingdiary.database.sql.SqlAreaSektoren
import com.main.climbingdiary.database.sql.SqlRouten
import com.main.climbingdiary.database.sql.SqlRouten.getInsertProjektTasks
import com.main.climbingdiary.database.sql.SqlRouten.getInsertRouteTasks
import com.main.climbingdiary.database.sql.SqlRouten.getRouteList
import com.main.climbingdiary.database.sql.SqlStatistic
import com.main.climbingdiary.models.RouteType
import java.nio.charset.StandardCharsets

object TaskRepository {
    private val TAG = "DataAdapter"
    private val mDb: SQLiteDatabase
    private val context: Context by lazy { MainActivity.getMainAppContext() }

    init {
        val mDbHelper = DatabaseHelper(context, null)
        mDbHelper.openDataBase()
        mDb = mDbHelper.writableDatabase
    }

    fun getAllRoutes(): Cursor {
        return getCursor(getRouteList(RouteType.ROUTE))
    }

    fun getAllProjekts(): Cursor {
        return getCursor(getRouteList(RouteType.PROJEKT))
    }

    fun getRoute(id: Int): Cursor {
        return getCursor(SqlRouten.getRoute(id))
    }

    fun getProjekt(id: Int): Cursor {
        return getCursor(SqlRouten.getProjekt(id))
    }

    fun getAllAreas(): Cursor {
        return getCursor(SqlAreaSektoren.getAllAreas())
    }

    fun getSectorListByAreaName(name: String): Cursor {
        return getCursor(SqlAreaSektoren.getSectorByAreaName(name))
    }

    fun getSectorIdByAreaNameAndSectorName(sectorName: String?, areaName: String?): Cursor {
        return getCursor(SqlAreaSektoren.getSectorIdBySectorNameAndAreaName(sectorName, areaName))
    }

    fun getAreaIdByAreaNameAndSectorName(sectorName: String?, areaName: String?): Cursor {
        return getCursor(SqlAreaSektoren.getAreaIdBySectorNameAndAreaName(sectorName, areaName))
    }

    fun getAreaByAreaName(areaName: String?): Cursor {
        return getCursor(SqlAreaSektoren.getAreaByName(areaName))
    }

    fun getTopTenRoutes(year: Int): Cursor {
        return getCursor(SqlRouten.getTopTenRoutes(year))
    }

    fun getYears(filterSet: Boolean): Cursor {
        return getCursor(SqlRouten.getYears(filterSet))
    }

    fun getTableValues(): Cursor {
        return getCursor(SqlStatistic.getTableValues())
    }

    fun getBarChartValues(): Cursor {
        return getCursor(SqlStatistic.getBarChartValues())
    }

    fun insertRoute(route: Route): Boolean {
        //create the transaction
        return executeSqlTasks(getInsertRouteTasks(route))
    }

    fun insertArea(area: Area): Boolean {
        return executeSqlTasks(arrayOf(SqlAreaSektoren.insertArea(area)))
    }

    fun insertSector(sector: Sector): Boolean {
        return executeSqlTasks(arrayOf(SqlAreaSektoren.insertSector(sector)))
    }

    fun insertProjekt(projekt: Projekt): Boolean {
        return executeSqlTasks(getInsertProjektTasks(projekt))
    }

    fun updateRoute(route: Route): Boolean {
        return executeSqlTasks(arrayOf(SqlRouten.updateRoute(route)))
    }

    fun deleteRoute(id: Int): Boolean {
        return executeSqlTasks(arrayOf(SqlRouten.deleteRoute(id)))
    }

    fun deleteProjekt(id: Int): Boolean {
        return executeSqlTasks(arrayOf(SqlRouten.deleteProjekt(id)))
    }

    fun cleanDatabase():Boolean{
        val tasks = arrayOf(
            "DELETE FROM routen_bouldern",
            "DELETE FROM routen_klettern",
            "DELETE FROM projekte_bouldern",
            "DELETE FROM projekte_klettern",
            "DELETE FROM gebiete_klettern",
            "DELETE FROM gebiete_bouldern",
            "DELETE FROM sektoren_bouldern",
            "DELETE FROM sektoren_klettern"
        )
        return executeSqlTasks(tasks)
    }

    @SuppressLint("Recycle")
    fun getCursor(sql: String): Cursor {
        return try {
            val mCur = mDb.rawQuery(sql, null)
            mCur.moveToNext()
            mCur
        } catch (mSQLExeption: SQLException) {
            Log.e(TAG, "Error >>$mSQLExeption")
            throw mSQLExeption
        }
    }

    fun executeSqlTasks(tasks: Array<String>): Boolean {
        mDb.beginTransaction()
        return try {
            for (x in tasks) {
                Log.d("Execute", x)
                val query = x.trimIndent()
                mDb.execSQL(String(query.toByteArray(), StandardCharsets.UTF_8))
            }
            mDb.setTransactionSuccessful()
            true
        } catch (exception: SQLiteException) {
            Log.e(TAG, "Error >>$exception")
            false
        } finally {
            mDb.endTransaction()
        }
    }
}